### PR TITLE
feat: add notify-about-build-failure GHA

### PR DESCRIPTION
**Pull request by sashuk**
[FX-4324]

### Description

This pull request adds reusable `notify-about-build-failure` action.

### How to test

1. Restart the https://github.com/toptal/staff-portal/actions/runs/6460088976 job (it uses latest version of the GHA from this branch), notifications should arrive to https://toptal-core.slack.com/archives/C05VBSKHN5U and personal messages of @sashuk 

### Screenshots

The notifications below are sent with the following settings (tested in https://github.com/toptal/staff-portal/pull/11180):

```
slack-bot-token: ${{ secrets.SLACK_BOT_TOKEN }}
top-team-api-key: ${{ secrets.TOP_TEAM_API_KEY }}
slack-channel-name: '#-sashuk-test-notifications-channel'
github-commit-author-name: ${{ github.event.pusher.name || 'sashuk' }}
github-action-run-url: https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}
```

---

Slack direct messages

Screenshot 2023-10-09 at 18 01 01

---

Slack channel

Screenshot 2023-10-09 at 18 01 15

### Development checks

- [x] Add changeset according to [guidelines](https://github.com/toptal/davinci-github-actions/blob/master/_docs/contribution/changeset-guidelines.md) (if needed)

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-anvil ping reviewers` - Ping FX team for review

</details>

[FX-4324]: https://toptal-core.atlassian.net/browse/FX-4324?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by sashuk** [comment]
@toptal-anvil ping reviewers

File: package.json [hunk 1]
```diff
@@ -13,7 +13,8 @@
     "prepare": "husky install",
     "build:create-matrix": "ncc build create-matrix/index.js -o create-matrix/dist",
     "build:get-changeset-info": "ncc build get-changeset-info/index.js -o get-changeset-info/dist",
-    "build:report-missing-changeset": "ncc build report-missing-changeset/index.js -o report-missing-changeset/dist"
+    "build:report-missing-changeset": "ncc build report-missing-changeset/index.js -o report-missing-changeset/dist",
+    "build:notify-about-build-failure": "ncc build notify-about-build-failure/index.js -o notify-about-build-failure/dist"
```
Comment (augustobmoura) [inline package.json:17]: Do we really need a build step for these packages? What are the benefits of doing this?

Comment (sashuk) [reply package.json:17]: @augustobmoura We need them due to imports like `@actions/core`, they need to be baked into the JS file (this was the problem I've raised on last Daily, it turned out this compilation step was needed)

**Comment by sashuk** [comment]
@toptal-anvil ping reviewers

File: .changeset/little-cows-tap.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'davinci-github-actions': minor
+---
+
+- add notify-about-build-failure action
```
Comment (dmaklygin) [inline .changeset/little-cows-tap.md:5]: I would add two more words about it in the changeset

File: .lintstagedrc.js [hunk 3]
```diff
@@ -7,16 +9,12 @@ module.exports = {
...
-    'git add report-missing-changeset/dist',
-  ],
+  ...(targets.map((acc, target) => {
+    acc[`${target}/**/*.js`] = () => [
+      `yarn build:${target}`,
+      `git add ${target}/dist`,
+    ]
+
+    return acc
+  }, {}))
```
Comment (dmaklygin) [inline .lintstagedrc.js:19]: maybe we should generalize building?
do not use targets, but find `action.js` in the action's folder and apply `yarn build:package`?

File: package.json (hunk 1 above)
Comment (dmaklygin) [reply package.json:17]: I would actually generalize building process, to have only a single command

File: .changeset/little-cows-tap.md (hunk 2 above)
Comment (sashuk) [reply .changeset/little-cows-tap.md]: @dmaklygin Extended, thanks!

File: .lintstagedrc.js (hunk 3 above)
Comment (sashuk) [reply .lintstagedrc.js:19]: @dmaklygin I've removed the corresponding `scripts` from `package.json` as they are not needed, but the `some-action/index.js` detection could be probably added in a separate ticket?

**Comment by sashuk** [comment]
@toptal-anvil ping reviewers

**Review by dmaklygin — APPROVED** [review]

File: report-missing-changeset/index.js [hunk 4]
```diff
@@ -19,6 +14,7 @@ console.log('Release packages:', releasePackages)
 console.log('Changed packages:', changedPackages)
 console.log('Missing packages:', missingPackages)
 
+const octokit = new github.getOctokit(GITHUB_TOKEN)
```
Comment (mkrl) [inline report-missing-changeset/index.js:17]: I'm curious how this was available before, is `octokit` available in all modules in GH actions runner and it changed? Was it previouisly broken?

**Review by mkrl — APPROVED** [review]
Looks good! I have one nit on wording and a question :smile:

File: report-missing-changeset/index.js (hunk 4 above)
Comment (sashuk) [reply report-missing-changeset/index.js:17]: @mkrl Oh my bet, I'll revert this change, it was just testing 🙂

File: notify-about-build-failure/index.js [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    'Content-Type': 'application/json',
+    Authorization: `Bearer ${slackBotToken}`,
+  },
+}
+
+const getSlackMessage = (
+  slackHandle,
+  githubActionRunUrl
+) => `Hello <@${slackHandle}>!
+Build ${githubActionRunUrl} is currently failing on master, please fix to unblock or let the proper team know.`
```
[the remark this answers is not in the document]
Comment (sashuk) [reply notify-about-build-failure/index.js]: @mkrl Fixed, thanks!

**Pull request merged**